File: 2_Inventory_Model/1_Model/Optimizer_Case/Optimizer_Markov.m

clear all
close all hidden
format long
clc
addpath('src');

%_ _SOLUTION METHOD _ _ _ _ _ _ _  _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
solution_method='R2';              %exact, CEC, R1, R2
%_ _SUBOPTIMAL DECISIONS _ _ _ _ _ _ _  _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
suboptimal_decision_vector_R1=xlsread('Stochastic_Regimes_MR-MO.xlsx',1,'B4:B11');
suboptimal_decision_vector_R2=xlsread('Stochastic_Regimes_MR-MO.xlsx',1,'C4:C11');
suboptimal_decision_matrix_CEC=xlsread('Stochastic_Regimes_MR-MO.xlsx',1,'F4:M11');
%_ _INVENTORY COST _ _ _ _ _ _ _  _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
variables.t_max=4;                       
variables.ch=17.5;                            
variables.cp=6000;  
%_ _DEMAND _ _ _ _ _ _ _  _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
variables.d_max=1;                        
variables.I_max=4;                       
demand_dist='manual';                 %uniform, normal, poisson, nbin, manual
    variables.normal.mu=15;
    variables.normal.sigma=3;
    variables.poisson.lambda=1;
    variables.nbin.r=22.5;
    variables.nbin.p=0.6;
start_inventory=0;    
%_ _PRICE PROCESS _ _ _ _ _ _ _  _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
variables.k=[0.99 0.01; 0.01 0.99];         
variables.m=2;                              
beliefs=[0.5];
variables.prices=[1000 1500 2000 2500 3000 3500 4000 4500]; 
price_dist={'RW','MR'};
variables.P_pr_reg=0; %muss nur definiert sein, da unter variables, aber für Markov-Fall nicht von Bedeutung

%_ _CALCULATIONS START _ _ _ _ _ _ _  _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _    

diary myFirstDiaryFile.txt

for i=1:numel(variables.prices)
    observed_price=variables.prices(i);
    for h=1:1%numel(variables.prices)
        last_price=variables.prices(h);
        for j=1:1%numel(beliefs)
            pi1=beliefs(j);
            variables.pi=[pi1 1-pi1];
            
            suboptimal_decision_CEC=suboptimal_decision_matrix_CEC(j,i);
            suboptimal_decision_R1=suboptimal_decision_vector_R1(i);
            suboptimal_decision_R2=suboptimal_decision_vector_R2(i);


            allSave=cell(variables.t_max, 1);           



            for T=variables.t_max 
                variables.t_max=T;
            tic
                periods=backward_recursion_markov(variables,observed_price,last_price,start_inventory,demand_dist,price_dist,solution_method,suboptimal_decision_CEC,suboptimal_decision_R1,suboptimal_decision_R2);
                if ~isempty(periods)
                disp(['<------------Planning horizon : ' num2str(T) ' periods --------------------->']);
                disp(['- - - - - - - - - - - - - - - - - - - - - - - - - - - - - -']);
                disp(['For planning horizon of ' num2str(T) ' periods, and observed price sequence: ' num2str(last_price) '-->' num2str(observed_price) ' and initial belief ' num2str(variables.pi) ':']);
                disp(['Optimal Order Decision: ' num2str(periods{1}.period_min_order)]);
                disp(['Minimum Total Expected Cost: ' num2str(periods{1}.period_min_cost)]);
                disp(['- - - - - - - - - - - - - - - - - - - - - - - - - - - - - -']);
                disp(['Suboptimal Cost (order decision (CEC): ' num2str(suboptimal_decision_CEC) '): ' num2str(periods{1}.suboptimal_cost_CEC)]);
                disp(['Optimality gap in % total cost (CEC): ' num2str((periods{1}.suboptimal_cost_CEC/periods{1}.period_min_cost)*100-100)]);
                disp(['- - - - - - - - - - - - - - - - - - - - - - - - - - - - - -']);
                disp(['Suboptimal Cost (order decision (R1): ' num2str(suboptimal_decision_R1) '): ' num2str(periods{1}.suboptimal_cost_R1)]);
                disp(['Optimality gap in % total cost (R1): ' num2str((periods{1}.suboptimal_cost_R1/periods{1}.period_min_cost)*100-100)]);
                disp(['- - - - - - - - - - - - - - - - - - - - - - - - - - - - - -']);
                disp(['Suboptimal Cost (order decision (R2): ' num2str(suboptimal_decision_R2) '): ' num2str(periods{1}.suboptimal_cost_R2)]);
                disp(['Optimality gap in % total cost (R2): ' num2str((periods{1}.suboptimal_cost_R2/periods{1}.period_min_cost)*100-100)]);
                end 
            toc
                allSave{T}={periods};
            end
        end 
        j+1;
    end 
    h+1;
end 
i+1;
diary off

File: 2_Inventory_Model/1_Model/Optimizer_Case/optimizer_markov.py
"""Inventory optimizer — Case Markov version.

Translated from Optimizer_Case/Optimizer_Markov.m.
Loads all parameters from a case data directory (default: 3_Case/Zinc/).
"""
import sys
import os
import time
import numpy as np

# Add parent Optimizer directory to path so we can import src_py
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'Optimizer'))
from src_py.backward_recursion_markov import backward_recursion_markov
from load_case import load_case_markov


def main(case_dir=None):
    if case_dir is None:
        case_dir = os.path.join(os.path.dirname(__file__), '..', '..',
                                '3_Case', 'Zinc')

    (variables, beliefs, suboptimal_decision_vector_R1,
     suboptimal_decision_vector_R2, suboptimal_decision_matrix_CEC,
     _, method_code, dist_code, start_inventory
     ) = load_case_markov(case_dir)

    prices = variables['prices']
    t_max = variables['t_max']
    # price_dist labels not needed — R1/R2 are loaded directly from files
    price_dist = ['R1', 'R2']

    # __ CALCULATIONS START __
    for i in range(len(prices)):
        observed_price = prices[i]
        for h in range(1):  # range(len(prices))
            last_price = prices[h]
            for j in range(1):  # range(len(beliefs))
                pi1 = beliefs[j]
                variables['pi'] = np.array([pi1, 1.0 - pi1])

                suboptimal_decision_CEC = int(suboptimal_decision_matrix_CEC[j, i])
                suboptimal_decision_R1 = int(suboptimal_decision_vector_R1[i])
                suboptimal_decision_R2 = int(suboptimal_decision_vector_R2[i])

                for T in [t_max]:
                    variables['t_max'] = T
                    tic = time.time()

                    periods = backward_recursion_markov(
                        variables, observed_price, last_price, start_inventory,
                        dist_code, price_dist, method_code,
                        suboptimal_decision_CEC, suboptimal_decision_R1,
                        suboptimal_decision_R2
                    )

                    toc = time.time()

                    if periods is not None:
                        print(f'<------------Planning horizon : {T} periods --------------------->')
                        print('- - - - - - - - - - - - - - - - - - - - - - - - - - - - - -')
                        print(f'For planning horizon of {T} periods, and observed price '
                              f'sequence: {last_price}-->{observed_price} and initial belief '
                              f'{variables["pi"]}:')
                        print(f'Optimal Order Decision: {periods[0]["period_min_order"]}')
                        print(f'Minimum Total Expected Cost: {periods[0]["period_min_cost"]}')
                        print('- - - - - - - - - - - - - - - - - - - - - - - - - - - - - -')
                        print(f'Suboptimal Cost (order decision (CEC): '
                              f'{suboptimal_decision_CEC}): '
                              f'{periods[0]["suboptimal_cost_CEC"]}')
                        opt_gap_CEC = (periods[0]['suboptimal_cost_CEC']
                                       / periods[0]['period_min_cost'][0] * 100 - 100)
                        print(f'Optimality gap in % total cost (CEC): {opt_gap_CEC}')
                        print('- - - - - - - - - - - - - - - - - - - - - - - - - - - - - -')
                        print(f'Suboptimal Cost (order decision (R1): '
                              f'{suboptimal_decision_R1}): '
                              f'{periods[0]["suboptimal_cost_R1"]}')
                        opt_gap_R1 = (periods[0]['suboptimal_cost_R1']
                                      / periods[0]['period_min_cost'][0] * 100 - 100)
                        print(f'Optimality gap in % total cost (R1): {opt_gap_R1}')
                        print('- - - - - - - - - - - - - - - - - - - - - - - - - - - - - -')
                        print(f'Suboptimal Cost (order decision (R2): '
                              f'{suboptimal_decision_R2}): '
                              f'{periods[0]["suboptimal_cost_R2"]}')
                        opt_gap_R2 = (periods[0]['suboptimal_cost_R2']
                                      / periods[0]['period_min_cost'][0] * 100 - 100)
                        print(f'Optimality gap in % total cost (R2): {opt_gap_R2}')

                    print(f'Elapsed time: {toc - tic:.3f} seconds')


if __name__ == '__main__':
    case = sys.argv[1] if len(sys.argv) > 1 else None
    main(case)
